matgen/generate_matrix_he_and_tz.cc

```diff
@@ -0,0 +1,323 @@
+                    for (int64_t i = i_start; i < i_end; ++i) {
+                        if (A.tileIsLocal( i, j )) {
+                            #pragma omp task slate_omp_default_none shared( A ) \
+                                firstprivate( i, j, jj, zero )
+                            {
+                                A.tileGetForWriting( i, j, LayoutConvert::ColMajor );
+                                auto Aij = A( i, j );
+                                lapack::laset(
+                                    lapack::MatrixType::General,
+                                    Aij.mb(), 1, zero, zero,
+                                    &Aij.at( 0, jj ), Aij.stride() );
+                            }
+                        }
+                    }
+                }
+                j_global += A.tileNb( j );
+            }
+        }
+
+        // Set row A( zero_col, : ) = 0.
+        // Merging this with the above omp parallel would require
+        // todo: this should apply only to Hermitian or symmetric matrices,
+        // not to triangular or trapezoid matrices. Perhaps
+        //     generate_matrix( ..., HermitianMatrix, ... )
+        // could set a flag when calling this routine.
+        #pragma omp parallel
+        #pragma omp master
+        {
+            int64_t i_global = 0;
+            for (int64_t i = 0; i < mt; ++i) {
+                int64_t ii = zero_col - i_global;
+                if (0 <= ii && ii < A.tileMb( i )) {
+                    // upper trapezoid is j in [ i, nt )
+                    // lower trapezoid is j in [ 0, min( i+1, nt ) )
+                    int64_t j_start = A.uplo() == Uplo::Upper ? i  : 0;
+                    int64_t j_end   = A.uplo() == Uplo::Upper ? nt : std::min( i+1, nt );
+                    for (int64_t j = j_start; j < j_end; ++j) {
+                        if (A.tileIsLocal( i, j )) {
+                            #pragma omp task slate_omp_default_none shared( A ) \
+                                firstprivate( i, j, ii, zero )
+                            {
+                                A.tileGetForWriting( i, j, LayoutConvert::ColMajor );
+                                auto Aij = A( i, j );
+                                lapack::laset(
+                                    lapack::MatrixType::General,
+                                    1, Aij.nb(), zero, zero,
+                                    &Aij.at( ii, 0 ), Aij.stride() );
+                            }
+                        }
+                    }
+                }
+                i_global += A.tileMb( i );
+            }
+        }
+    }
+
+    A.tileUpdateAllOrigin();
+}
+
+// -----------------------------------------------------------------------------
+/// Generates an m-by-n Hermitian-storage test matrix.
+/// Handles Hermitian matrices.
+/// Diagonal elements of a Hermitian matrix must be real;
+/// their imaginary part must be 0.
```

Comment:
Is this true? In LAPACK, the imaginary part is just ignored.

Reply:
While most routines should ignore it, SLATE doesn't always ignore it.  For example, `hettmqr` doesn't ignore the imaginary component of the diagonal elements.  [It converts the diagonal tiles to full storage without touching the diagonal elements.](https://github.com/icl-utk-edu/slate/blob/master/src/internal/internal_hettmqr.cc#L190)  I don't know if there are any other such routines, but I'd suggest auditing the code before making that assumption.